[WebAssembly] Add ifdefs for the WASI target

MaxDesiatov opened the pull request:
<!-- What's in this pull request? -->
Disables/enables stdlib and IRGen code appropriately when targeting WebAssembly/WASI.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
This is a part of [SR-9307](https://bugs.swift.org/browse/SR-9307) and #24684.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

(cc @kateinoigakukun)

MaxDesiatov committed:
[WebAssembly] Add ifdefs for the WASI target

compnerd commented on include/swift/SwiftRemoteMirror/Platform.h:
```diff
@@ -18,7 +18,7 @@ extern "C" {
 
 #if defined(swiftRemoteMirror_EXPORTS)
-# if defined(__ELF__)
+# if defined(__ELF__) || defined(__wasi__)
```
I think that this should be `__WASM__` as the object file format is WASM, not WASI.

compnerd commented on include/swift/SwiftRemoteMirror/Platform.h:
```diff
@@ -30,9 +30,7 @@ extern "C" {
-# if defined(__ELF__)
-#   define SWIFT_REMOTE_MIRROR_LINKAGE __attribute__((__visibility__("default")))
-# elif defined(__MACH__)
+# if defined(__ELF__) || defined(__MACH__) || defined(__wasi__)
```
Similar

compnerd commented on lib/IRGen/IRGen.cpp:
```diff
@@ -165,6 +165,12 @@ swift::getIRTargetOptions(const IRGenOptions &Opts, ASTContext &Ctx) {
   auto *Clang = static_cast<ClangImporter *>(Ctx.getClangModuleLoader());
+
+  // WebAssembly doesn't support atomics yet, see https://bugs.swift.org/browse/SR-12097
+  if (Clang->getTargetInfo().getTriple().isOSBinFormatWasm()) {
```
Remove the extraneous braces.  IIRC, we set the `ThreadModel` for cygwin/MinGW as well.  Can we please set the thread model at one site?

MaxDesiatov replied:
As far as I'm aware this is a single place where `ThreadModel` is used in the whole Swift project, it's not set for Cygwin anywhere. The only somewhat related settings for Cygwin that I see are set in `createTargetMachine`, but that doesn't give access to `TargetOpts`:

```swift
  if (EffectiveTriple.isArch64Bit() && EffectiveTriple.isWindowsCygwinEnvironment())
    cmodel = CodeModel::Large;
```

compnerd commented on stdlib/private/StdlibUnittest/InterceptTraps.cpp:
```diff
@@ -26,6 +26,9 @@
 #include "swift/Runtime/Config.h"
 
+// no signals support on WASI yet, see https://github.com/WebAssembly/WASI/issues/166
+#if !defined(__wasi__)
```
I think that it's better to preprocess away the entire file.  I don't remember there being any symbols referenced from here.

compnerd commented on stdlib/private/StdlibUnittest/RaceTest.swift:
```diff
@@ -562,7 +562,12 @@ class _InterruptibleSleep {
 
+    #if os(WASI)
+    // WebAssembly/WASI on wasm32 is the only 32-bit platform with Int64 time_t
+    var timeout = timeval(tv_sec: time_t(duration), tv_usec: 0)
```
Is it not safe to always cast the `duration` to `time_t`?

MaxDesiatov replied:
`duration` is `Int` here, which is 32-bit on wasm32, do you mean a force cast?

```swift
duration as? time_t
```

Would that be safe then?

compnerd commented on stdlib/private/StdlibUnittest/StdlibUnittest.swift:
```diff
@@ -1447,7 +1450,12 @@ public func runAllTests() {
   } else {
+    #if os(WASI)
+    // WASI doesn't support child processes
+    var runTestsInProcess: Bool = true
```
This won't work.  The death tests need to be run as child processes.  I think it is better to not make the test suite harness more complicated for something which won't work.  That is drop the changes to `stdlib/private/StdlibUnittest`

compnerd commented on stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:
```diff
@@ -13,14 +13,19 @@
 internal func _signalToString(_ signal: Int) -> String {
+  #if os(WASI)
+  // no signals support on WASI yet, see https://github.com/WebAssembly/WASI/issues/166
+  fatalError("\(#function) is not supported on WebAssembly/WASI")
```
I don't think that you should stub out `_signalToString`.  Instead, you need to die earlier at the point where the signal is being handled.  Just drop the entire function on WASI.

compnerd commented on stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:
```diff
@@ -141,6 +146,15 @@ public func waitProcess(_ process: HANDLE) -> ProcessTerminationStatus {
+// WASI doesn't support child processes
+public func spawnChild(_ args: [String])
+  -> (pid: pid_t, stdinFD: CInt, stdoutFD: CInt, stderrFD: CInt) {
+  preconditionFailure("\(#function) is not supported on WebAssembly/WASI")
```
This should be `fatalError`, not `preconditionFailure`.

compnerd commented on stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:
```diff
@@ -141,6 +146,15 @@ public func waitProcess(_ process: HANDLE) -> ProcessTerminationStatus {
+  preconditionFailure("\(#function) is not supported on WebAssembly/WASI")
+}
+public func posixWaitpid(_ pid: pid_t) -> ProcessTerminationStatus {
+  preconditionFailure("\(#function) is not supported on WebAssembly/WASI")
```
Similar

compnerd commented on stdlib/private/SwiftPrivateLibcExtras/SwiftPrivateLibcExtras.swift:
```diff
@@ -125,6 +125,8 @@ public func _stdlib_pipe() -> (readEnd: CInt, writeEnd: CInt, error: CInt) {
 #if os(Windows)
     return _pipe(unsafeFds.baseAddress, 0, 0)
+#elseif os(WASI)
+    preconditionFailure("no pipes available on WebAssembly/WASI")
```
IIRC, this is largely used for the reflection tests.  Perhaps we should just disable the reflection tests instead.  I really think that given the amount of stuff not available in WASI libc yet, the focus should be improving that first (which is why I stopped and started looking at WASI libc instead).

MaxDesiatov replied:
This is used in `RaceTest.swift` in the implementation of `_InterruptibleSleep` on non-Windows platforms. Would `_InterruptibleSleep` need a dedicated implementation for WASI instead? Could WebAssembly/WASI#77 be somewhat related?

compnerd commented on stdlib/private/SwiftPrivateThreadExtras/ThreadBarriers.swift:
```diff
@@ -12,7 +12,7 @@
 #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
 import Darwin
-#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
+#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku) || os(WASI)
```
I think that it might be better to just not build `SwiftPrivateThreadExtras` on WASI

MaxDesiatov replied:
This would make `RaceTest` that uses `SwiftPrivateThreadExtras` disabled, requiring us to disable quite a few tests. Do you think that be worth it?

gribozavr replied:
+1 to disabling the whole module, if it leads to simpler code. Race tests would not work (and are pointless) without threads anyway.

compnerd reviewed (changes requested):
This is a huge change.  Please split this up into more focused changes.  There can be a single change for the runtime, and a single change for the standard library at least.

MaxDesiatov commented on stdlib/private/SwiftPrivateThreadExtras/ThreadBarriers.swift:
```diff
@@ -35,7 +35,9 @@ public struct _stdlib_thread_barrier_t {
-  var cond: UnsafeMutablePointer<pthread_cond_t?>?
+  var cond: UnsafeMutablePointer<pthread_cond_t?>
+#elseif os(WASI)
+  // WASI environment has a only single thread, see https://bugs.swift.org/browse/SR-12097
```
Does this comment refer to some function? It's currently positioned within a struct property declaration, do you mean that the structs `init` should invoke `fatalError` or `preconditionFailure` when it runs on WASI?

gribozavr replied:
Sorry about sloppy commenting -- yes, I was trying to attach the comment to the `init`, and invoking `fatalError` is exactly what I meant.

kateinoigakukun commented:
@MaxDesiatov Could you reflect this change also for WASI? https://github.com/apple/swift/pull/29536
This prevents from building WASI target stdlib

gribozavr commented on stdlib/private/StdlibUnittest/SymbolLookup.swift:
```diff
@@ -23,7 +23,7 @@
   let RTLD_DEFAULT = UnsafeMutableRawPointer(bitPattern: -2)
-#elseif os(Linux)
+#elseif os(Linux) || os(WASI)
   let RTLD_DEFAULT = UnsafeMutableRawPointer(bitPattern: 0)
```
WebAssembly does not have dynamic linking, does it? If it does not, then `RTLD_DEFAULT` is not meaningful I think.

gribozavr commented on stdlib/private/SwiftPrivateLibcExtras/Subprocess.c:
```diff
@@ -11,7 +11,7 @@
 //===----------------------------------------------------------------------===//
 
 // posix_spawn is not available on Android or Windows (MSVC).
```
Please update the comment.

gribozavr commented on stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:
```diff
@@ -44,7 +47,12 @@ public enum ProcessTerminationStatus : CustomStringConvertible {
     case .exit(let status):
       return "Exit(\(status))"
     case .signal(let signal):
+      #if os(WASI)
```
Indent -8? Seems like other `#if`s start in the first column.

gribozavr commented on stdlib/public/Platform/Platform.swift:
```diff
@@ -380,7 +382,7 @@ public var SEM_FAILED: UnsafeMutablePointer<sem_t>? {
   // The value is ABI.  Value verified to be correct for OS X, iOS, watchOS, tvOS.
   return UnsafeMutablePointer<sem_t>(bitPattern: -1)
-#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
+#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku) || os(WASI)
```
Does `sem_open()` exist and work on WASI? If not, ifdef out the whole variable.

MaxDesiatov replied:
It does exist and `semaphore.h` is available too.

gribozavr commented on stdlib/public/core/AtomicInt.swift.gyb:
```diff
@@ -65,7 +65,7 @@ internal func _swift_stdlib_atomicCompareExchangeStrongInt(
   expected: UnsafeMutablePointer<Int>,
   desired: Int) -> Bool {
-#if arch(i386) || arch(arm)
+#if arch(i386) || arch(arm) || arch(wasm32)
```
Other comments that you left say that WebAssembly has no atomics right now. So how do these functions get compiled? Regular loads and stores?

MaxDesiatov replied:
Yes, this is set in `IRGen.cpp` change in this same PR:

```swift
  if (Clang->getTargetInfo().getTriple().isOSBinFormatWasm())
    TargetOpts.ThreadModel = llvm::ThreadModel::Single;
```

MaxDesiatov commented:
Hi @gribozavr @compnerd I've addressed your feedback and added comments with clarifications, could you have another look. Thank you!

gribozavr commented:
I don't see where in the patch you disable building `SwiftPrivateThreadExtras` on WASI...

If you want to submit this change as an incremental improvement, LGTM, but just noting that it seems that some work is still needed to disable building some modules.

MaxDesiatov commented:
@gribozavr in ` stdlib/private/StdlibUnittest/CMakeLists.txt` there's no `SWIFT_MODULE_DEPENDS_WASI` yet, but when it's there it won't specify `SwiftPrivateThreadExtras` to exclude that dependency. `SWIFT_MODULE_DEPENDS_WASI` implementation would come in a future PR that updates the build system and CMake code.

Would you mind triggering CI for this PR? Thanks again for your help.

gribozavr commented:
@MaxDesiatov That will avoid linking the `StdlibUnittest` against `SwiftPrivateThreadExtras`, but still won't exclude `SwiftPrivateThreadExtras` from being built, I believe. PTAL at `TARGET_SDKS` as used in `stdlib/public/Platform/CMakeLists.txt`, for example.

gribozavr commented:
@swift-ci Please test

MaxDesiatov commented:
@gribozavr I've updated `CMakeLists.txt` to use the `TARGET_SDKS` argument instead of `SWIFT_MODULE_DEPENDS`, also rebased the PR on top of `master`, as the CI issue seems to be caused by the branch being too out of sync:

```
swiftpm/Sources/PackageLoading/Diagnostics.swift:11:8: error: 
compiled module was created by a newer version of the compiler: 
buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/tsc/swift/TSCBasic.swiftmodule
```

Could trigger another CI run? Thanks

futurejones commented:
Can one of the admins verify this patch?

gribozavr commented:
I think we need both `TARGET_SDKS` (to prevent the module from being built when you build everything that's available), and `SWIFT_MODULE_DEPENDS` (to prevent other modules from depending on it when it is not available.)

gribozavr commented:
@swift-ci Please test

MaxDesiatov commented:
@gribozavr yeah, it does have both, I've restored `SWIFT_MODULE_DEPENDS` to its original state. As far as I understand, when `SWIFT_MODULE_DEPENDS_WASI` is available we'd just add that with a corresponding setting, while `TARGET_SDKS` should be enough for now to avoid building it for WASI.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please smoke test

MaxDesiatov commented:
Thanks for triggering CI! As it has passed and PR got the approval, can it be merged now to avoid potential conflicts appearing and preventing it from being merged as the diff is big enough?

stephentyrone commented:
In `StringObject.swift`, you seem to be missing `arch(wasm32)` in the very first conditional, defining the string layout. If that's the case, am I correct in believing that none of this is actually tested for wasm?

MaxDesiatov commented:
Do you mean line 80 that precedes `internal enum Variant`? That one has a corresponding `#else` which does the job for other platforms in my understanding, but I can't confirm that it works for WASI at the moment. Will need to have a closer look if some of our failing tests have any relation to it, we are slowly enabling the test suite here in swiftwasm/swift#12 and swiftwasm/swift#30. If it does fail, I'd be happy to fix it in a follow-up PR.

stephentyrone commented:
Yes, that one; the fields of a `_StringObject` are radically different on 32-bit platforms than they are on 64-bit platforms; if you have the wrong definition (as I think you do), it shouldn't even build.

MaxDesiatov commented:
Well, it does build it seems, but thanks for having a look, I'll keep that in mind 👍
